### DISCUSSION TITLE
MTConnect-Agent_MQTT-Relay-Gateway with mutliple machines

### DISCUSSION THREAD

**Discussion by DouglasRudd** [TrakHound/MTConnect.NET#52] [Q&A] [answered]
I have tried looking for and following any documentation related to connecting to multiple MTConnect ( I have 2 new Haas VMCs).
I have created a folder called "Devices" in which I have placed a copy of the XML from the probe results for each machine.
I then added the following information in the agent.config.yaml 

 HTTP Client Adapter Configuration -

devices: Devices

clients:
- address: 10.241.188.206
  port: 8082
  deviceKey: VF-5_50
- address: 10.241.188.205
  port: 8082
  deviceKey: VF-10_50

when I start the agent sometimes it will update both machine mqtt data, but 90% of the time it will only update the VF-10_50, showing the VF-5_50 as unavailable. I have issued an http sample, and current corren request in a browser to verify the machine is available, but the mqtt topic its still showing unavailable.

what am I missing or doing wrong?
do I need to run two instances of the application one for each machine?

thanks

**Comment by PatrickRitchie** [comment, marked as answer]
You should be able to read from two separate external agents with the same instance of the application.

It sounds like you have it setup correctly though you may want to double check:

- The DeviceKey is set the same as the XML probe in the Devices directory
- The Probe XML only contains the Device you are targeting
- You may want to look at the logs to see if the client is disconnecting. You can use the Heartbeat property in the client configuration to adjust this (some agents don't support a heartbeat so you have to set it to "0")
- I have heard before that the Haas Agent either has trouble with or doesn't support streaming. If this is the case, you can set the CurrentOnly property in the client configuration and instead of streaming it will use polling

Let me know if any of the above works for you or if you have any other questions.

**Reply by DouglasRudd** [reply]
after a bit of trial and error. I have managed to get the feeds to work for both machines.
I did set hearbeat = 0.
the client logs are showing the following.

2023-11-09 08:58:14.2633|0|INFO|client-logger|[Http-Adapter] : ID = adapter_http_cb9e1e45c0 : Client Stream Started : http://10.241.188.205:8082/VF-10_50/sample?count=1000 
2023-11-09 08:58:15.8395|0|INFO|client-logger|[Http-Adapter] : ID = adapter_http_97eeb9528a : Client Stream Started : http://10.241.188.206:8082/VF-5_50/sample?count=1000 
2023-11-09 08:58:33.1665|0|INFO|client-logger|[Http-Adapter] : ID = adapter_http_cb9e1e45c0 : Client Stream Stopped 
2023-11-09 08:58:34.7435|0|INFO|client-logger|[Http-Adapter] : ID = adapter_http_97eeb9528a : Client Stream Stopped 

is this the expected behavior?
I also had to set the interval = -500 (Haas does not use intervals and returns an error)
is there a way to customize the mqtt topics? I would like to format it to match the current UNS I have set up.

**Reply by PatrickRitchie** [reply]
I think I may have told you wrong about the CurrentOnly property. It looks like the **UseStreaming** property may fit your scenario better and it is set in the client configuration: https://github.com/TrakHound/MTConnect.NET/tree/master/src/MTConnect.NET-HTTP/Clients#properties

```yaml
clients:
- address: 10.241.188.206
  port: 8082
  deviceKey: VF-5_50
  interval: 5000
  useStreaming: false
```

Note that this will only return a "snapshot" and not the observations between the interval. I think I should be able to accommodate the scenario you have by adding a sample request at the polling interval if needed. You can see the polling logic here: https://github.com/TrakHound/MTConnect.NET/blob/master/src/MTConnect.NET-HTTP/Clients/MTConnectHttpClient.cs#L705

For MQTT topics, you can use the TopicPrefix property in the mqtt configuration. Let me know how that works for you and if you need additional functionality as I'm looking for feedback on how to develop that further. I'm also interested to see how well this format integrates with a UNS as you stated.
